tests/integration_test/service_control_quota_test/service_control_quota_test.go

```diff
@@ -249,19 +249,156 @@ func TestServiceControlQuotaUnavailable(t *testing.T) {
 		wantScRequestCount    int
 		wantScRequests        []interface{}
 		wantGetScRequestError error
+	}{
+		{
+			desc:           "first request is granted with 3 service calls: check, quota and report",
+			clientProtocol: "http",
+			method:         "/v1/shelves?key=api-key",
+			token:          testdata.FakeCloudTokenMultiAudiences,
+			httpMethod:     "GET",
+			wantResp:       `{"shelves":[{"id":"100","theme":"Kids"},{"id":"200","theme":"Classic"}]}`,
+			wantScRequests: []interface{}{
+				&utils.ExpectedCheck{
+					Version:         utils.ESPv2Version(),
+					ServiceName:     "bookstore.endpoints.cloudesf-testing.cloud.goog",
+					ServiceConfigID: "test-config-id",
+					ConsumerID:      "api_key:api-key",
+					OperationName:   "endpoints.examples.bookstore.Bookstore.ListShelves",
+					CallerIp:        platform.GetLoopbackAddress(),
+				},
+				&utils.ExpectedQuota{
+					ServiceName: "bookstore.endpoints.cloudesf-testing.cloud.goog",
+					MethodName:  "endpoints.examples.bookstore.Bookstore.ListShelves",
+					ConsumerID:  "api_key:api-key",
+					QuotaMetrics: map[string]int64{
+						"metrics_first":  2,
+						"metrics_second": 1,
+					},
+					QuotaMode:       scpb.QuotaOperation_BEST_EFFORT,
+					ServiceConfigID: "test-config-id",
+				},
+				&utils.ExpectedReport{
+					Version:                      utils.ESPv2Version(),
+					ServiceName:                  "bookstore.endpoints.cloudesf-testing.cloud.goog",
+					ServiceConfigID:              "test-config-id",
+					URL:                          "/v1/shelves?key=api-key",
+					ApiKeyInOperationAndLogEntry: "api-key",
+					ApiVersion:                   "1.0.0",
+					ApiKeyState:                  "VERIFIED",
+					ApiMethod:                    "endpoints.examples.bookstore.Bookstore.ListShelves",
+					ApiName:                      "endpoints.examples.bookstore.Bookstore",
+					ProducerProjectID:            "producer project",
+					ConsumerProjectID:            "123456",
+					FrontendProtocol:             "http",
+					BackendProtocol:              "grpc",
+					HttpMethod:                   "GET",
+					LogMessage:                   "endpoints.examples.bookstore.Bookstore.ListShelves is called",
+					StatusCode:                   "0",
+					// It always allow the first request, then cache its cost, accumulate all costs for 1 second,
+					// then call remote allocateQuota,  if fail, the next request will be failed with 429.
+					// Here is the first request.
+					ResponseCode: 200,
+					Platform:     util.GCE,
+					Location:     "test-zone",
+				},
+			},
+		},
+		{
+			// quota server return 404, with fail-open policy, cached quota result is positive.
```

Comment:
Add comment:

TODO(b/194517193): Consider fail close for client-side error (4xx errors)

Are you sure the quota server returning 404?

Reply:
Aded, yes, it is from fake quota sever at line 210